File: bin/client/typescript/Basic/BaseBinaryConfigData.ts
export class BaseBinaryConfigData
{
    private m_binary: Uint8Array = null;
    private m_index: number = 0;

    private m_valueMap: {[key: string]: any} = {};

    public parseData(binary: Uint8Array, fromIndex: number = 0): number
    {
        this.m_binary = binary;
        this.m_index = fromIndex;
        this.onParseData();
        this.m_binary = null;
        return this.m_index;
    }

    protected onParseData()
    {

    }

    protected setConfigValue(key: string, value: any)
    {
        this.m_valueMap[key] = value;
    }

    public getConfigValue(key: string): any
    {
        return this.m_valueMap[key];
    }

    public hasConfigValue(key: string): boolean
    {
        if(typeof(this.m_valueMap[key]) == "undefined")
        {
            return false;
        }
        return true;
    }

    protected readList(elementReader, target=this): any
    {
        let list = [];
        let count = this.readInt();
        for(let i = 0; i < count; i++)
        {
            list.push(elementReader.call(target));
        }
        return list;
    }

    protected readInt(): number
    {
        let b1: number = this.m_binary[this.m_index + 3];
        let b2: number = this.m_binary[this.m_index + 2];
        let b3: number = this.m_binary[this.m_index + 1];
        let b4: number = this.m_binary[this.m_index + 0];
        this.m_index += 4;
        let result = (b1 << (3 * 8)) + (b2 << (2 * 8)) + (b3 << (1 * 8)) + b4;
        return result;
    }

    protected readInt2(): any
    {
        let result = {
            x: this.readInt(),
            y: this.readInt(),
        };
        return result;
    }

    protected readInt3(): any
    {
        let result = {
            x: this.readInt(),
            y: this.readInt(),
            z: this.readInt(),
        };
        return result;
    }

    protected readInt4(): any
    {
        let result = {
            x: this.readInt(),
            y: this.readInt(),
            z: this.readInt(),
            w: this.readInt(),
        };
        return result;
    }

    protected readInt5(): any
    {
        let result = {
            x: this.readInt(),
            y: this.readInt(),
            z: this.readInt(),
            w: this.readInt(),
            v: this.readInt(),
        };
        return result;
    }

    protected readString(): string
    {
        let len = this.readInt();
        let bytes = this.readBytes(len);
        let str = "";
        let count = bytes.length;
        for(let i = 0; i < count; i++)
        {
            str += String.fromCharCode(bytes[i]);
        }
        let index = 0;
        let head = null;
        let body = null;
        let tail = null;
        count = str.length;
        for(let i = 0; i < count; i++)
        {
            index = str.indexOf("\\u");
            if(index == -1) break;
            head = index ? str.substring(0, index) : null;
            body = str.substring(index + 2, index + 2 + 4);
            tail = str.substring(index + 2 + 4);
            if(head)
            {
                str = head + String.fromCharCode(parseInt(body, 16)) + tail;
            }
            else
            {
                str = String.fromCharCode(parseInt(body, 16)) + tail;
            }
        }
        return str;
    }

    protected readBytes(len: number): number[]
    {
        let bytes: number[] = [];
        for(let i = 0; i < len; i++)
        {
            if(this.m_index >= this.m_binary.length)
            {
                bytes.push(0);
            }
            else
            {
                bytes.push(this.m_binary[this.m_index]);
                this.m_index++;
            }
        }
        return bytes;
    }
}

File: bin/client/csharp/Basic/BaseBinaryConfigData.cs
using System;
using System.Collections.Generic;

public class BaseBinaryConfigData
{
    private byte[] m_binary = null;
    private int m_index = 0;

    private Dictionary<string, object> m_valueMap = new Dictionary<string, object>();

    public void setData(byte[] binary)
    {
        m_binary = binary;
        onParseData();
    }

    protected void onParseData()
    {

    }

    protected void setConfigValue(string key, object value)
    {
        if (m_valueMap.ContainsKey(key))
        {
            m_valueMap[key] = value;
        }
        else
        {
            m_valueMap.Add(key, value);
        }
    }

    protected T getConfigValue<T>(string key)
    {
        if (m_valueMap.ContainsKey(key))
        {
            return (T)m_valueMap[key];
        }
        return default(T);
    }

    protected List<T> readList<T>(Func<T> elementReader)
    {
        List<T> list = new List<T>();
        int count = readInt();
        for (int i = 0; i < count; i++)
        {
            list.Add(elementReader());
        }
        return list;
    }

    protected int readInt()
    {
        int b1 = m_binary[m_index + 3];
        int b2 = m_binary[m_index + 2];
        int b3 = m_binary[m_index + 1];
        int b4 = m_binary[m_index + 0];
        m_index += 4;
        int result = (b1 << (3 * 8)) + (b2 << (2 * 8)) + (b3 << (1 * 8)) + b4;
        return result;
    }

    protected Int2 readInt2()
    {
        Int2 result = new Int2
        {
            x = readInt(),
            y = readInt()
        };
        return result;
    }

    protected Int3 readInt3()
    {
        Int3 result = new Int3
        {
            x = readInt(),
            y = readInt(),
            z = readInt()
        };
        return result;
    }

    protected Int4 readInt4()
    {
        Int4 result = new Int4
        {
            x = readInt(),
            y = readInt(),
            z = readInt(),
            w = readInt()
        };
        return result;
    }

    protected Int5 readInt5()
    {
        Int5 result = new Int5
        {
            x = readInt(),
            y = readInt(),
            z = readInt(),
            w = readInt(),
            v = readInt()
        };
        return result;
    }

    protected string readString()
    {
        int len = readInt();
        byte[] bytes = readBytes(len);
        string str = "";
        int count = bytes.Length;
        for (int i = 0; i < count; i++)
        {
            str += (char)bytes[i];
        }
        return str;
    }

    protected byte[] readBytes(int len)
    {
        byte[] bytes = new byte[len];
        for (int i = 0; i < len; i++)
        {
            if (m_index >= m_binary.Length)
            {
                bytes[i] = 0;
            }
            else
            {
                bytes[i] = m_binary[m_index];
                m_index++;
            }
        }
        return bytes;
    }
}

public class Int2
{
    public int x = 0;
    public int y = 0;
}

public class Int3 : Int2
{
    public int z = 0;
}

public class Int4 : Int3
{
    public int w = 0;
}

public class Int5 : Int4
{
    public int v = 0;
}

public class Int5 : Int5
{
    public int v = 0;
}